php_phongo.c

```diff
@@ -2459,7 +2459,138 @@ static char* php_phongo_manager_make_client_hash(const char* uri_string, zval* o
 	return hash;
 }
 
-static mongoc_client_t* php_phongo_make_mongo_client(const mongoc_uri_t* uri) /* {{{ */
+static bool php_phongo_extract_handshake_data(zval* zvalue, const char* key, char** value, size_t* value_len)
+{
+	if (Z_TYPE_P(zvalue) != IS_STRING) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"%s\" handshake option to be a string, %s given", key, PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zvalue));
+		return false;
+	}
+
+	*value     = estrdup(Z_STRVAL_P(zvalue));
+	*value_len = Z_STRLEN_P(zvalue);
+
+	return true;
+}
+
+static char* php_phongo_concat_handshake_data(const char* default_value, const char* custom_value, size_t custom_value_len)
+{
+	char*  ret;
+	size_t ret_len = strlen(default_value) + 1;
+
+	if (custom_value) {
+		ret_len += custom_value_len + 1;
+	}
+
+	ret = malloc(ret_len);
+
+	if (custom_value) {
+		snprintf(ret, ret_len, "%s|%s", default_value, custom_value);
+	} else {
+		snprintf(ret, ret_len, "%s", default_value);
+	}
+
+	return ret;
+}
+
+static void php_phongo_handshake_data_append(const char* name, size_t name_len, const char* version, size_t version_len, const char* platform, size_t platform_len)
+{
+	char* php_version_string;
+	char* driver_name;
+	char* driver_version;
+	char* full_platform;
+
+	php_version_string = malloc(4 + sizeof(PHP_VERSION));
+	sprintf(php_version_string, "PHP %s", PHP_VERSION);
+
+	driver_name    = php_phongo_concat_handshake_data("ext-mongodb:PHP", name, name_len);
+	driver_version = php_phongo_concat_handshake_data(PHP_MONGODB_VERSION, version, version_len);
+	full_platform  = php_phongo_concat_handshake_data(php_version_string, platform, platform_len);
+
+	MONGOC_DEBUG(
+		"Setting driver handshake data: name %s, version %s, platform %s",
+		driver_name,
+		driver_version,
+		full_platform);
+
+	mongoc_handshake_data_append(driver_name, driver_version, full_platform);
+
+	free(php_version_string);
+	free(driver_name);
+	free(driver_version);
+	free(full_platform);
+}
+
+static void php_phongo_set_handshake_data(zval* driverOptions)
+{
+	char*  name         = NULL;
+	size_t name_len     = 0;
+	char*  version      = NULL;
+	size_t version_len  = 0;
+	char*  platform     = NULL;
+	size_t platform_len = 0;
+
+	if (driverOptions && php_array_exists(driverOptions, "driver")) {
+		zval*      driver = php_array_fetch(driverOptions, "driver");
+		HashTable* ht_data;
+
+		if (Z_TYPE_P(driver) != IS_ARRAY && Z_TYPE_P(driver) != IS_OBJECT) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"driver\" driver option to be an array or object, %s given", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(driver));
+			return;
+		}
+
+		ht_data = HASH_OF(driver);
+
+		{
+			zend_string* string_key = NULL;
+			zend_ulong   num_key    = 0;
+			zval*        property;
+
+			ZEND_HASH_FOREACH_KEY_VAL_IND(ht_data, num_key, string_key, property)
```

Comment:
Any reason to use an iteration loop here instead of individual `php_array_fetch` lookups?

To contrast with a couple of other places that fully iterate through an options array:

 * `phongo_clientencryption_datakey_opts_from_zval` uses `ZEND_HASH_FOREACH_KEY_VAL` to check for non-string values and warn about extra keys.
 * `php_phongo_apply_options_to_uri` also does iteration through all keys because it needs to handle case variation.

I don't think this has any of those concerns, so I would have expected something more like what we do in `php_phongo_apply_driver_options_to_uri` or the Query constructor.

Reply:
The original idea was to have these handled case-insensitive as well, which isn't possible with the `php_array_exists` functions. Since other options (e.g. "context" are case-sensitive as well) I've changed this to no longer use a loop.